src/lib/sw_engine/tvgSwRaster.cpp

```diff
@@ -573,25 +627,30 @@ static bool _rasterImage(SwSurface* surface, uint32_t *img, uint32_t w, TVG_UNUS
     return true;
 }
 
-
 static bool _rasterImage(SwSurface* surface, const uint32_t *img, uint32_t w, uint32_t h, const SwBBox& region, const Matrix* invTransform)
 {
     for (auto y = region.min.y; y < region.max.y; ++y) {
         auto dst = &surface->buffer[y * surface->stride + region.min.x];
         auto ey1 = y * invTransform->e12 + invTransform->e13;
         auto ey2 = y * invTransform->e22 + invTransform->e23;
         for (auto x = region.min.x; x < region.max.x; ++x, ++dst) {
-            auto rX = static_cast<uint32_t>(roundf(x * invTransform->e11 + ey1));
-            auto rY = static_cast<uint32_t>(roundf(x * invTransform->e21 + ey2));
+            auto fX = x * invTransform->e11 + ey1;
+            auto fY = x * invTransform->e21 + ey2;
+            auto rX = static_cast<uint32_t>(roundf(fX));
+            auto rY = static_cast<uint32_t>(roundf(fY));
             if (rX >= w || rY >= h) continue;
-            auto src = img[rX + (rY * w)];    //TODO: need to use image's stride
+            uint32_t src;
+            if (rX == w - 1 || rY == h - 1 || (invTransform->e11 * invTransform->e11) + (invTransform->e21 * invTransform->e21) > 1) {
```

Comment:
@JSUYA I'm a bit worried about the performance. 
for each pixel we do this check and for most of the cases the first two conditions are not fulfilled, so the "invTransform..." condition is checked. 
Do you think, that it would be a good idea to check, whether it's a downscaling or not, a bit earlier? this way we would duplicate some code, I know... 
So it's rather a question than a suggestion. 

If you think it's pointless I'll accept and you can merge (or wait for other reviewers)

Reply:
Thanks for the review. Good point. I added a patch. If you have any better ideas please share me